Add initializers

JimClarke5 opened the pull request:
This PR adds initializers to `tensorflow-framework`. The classes inherit from `BaseInitializer` which implements `Initializer` which is also defined as a `Functional` interface so that an `Initializer` can be a lambda. Unit Test cases are also provided.

I have also included `utils.ShapeUtils`.   Some methods in `ShapeUtils`,

> ` public static boolean isCompatibleWith(Shape a, Shape b)`, 
> `public static Shape reduce(Shape shape, int axis) `
> `public static boolean isUnknownShape(Shape a) `

might be considered to be a part of  `org.tensorflow.ndarray.Shape`. The other methods bridge `Operands` to a `org.tensorflow.ndarray.Shape`, so probably don't belong in the `ndarray` module.  

This PR is not dependent on any other PR.

Craigacp commented on tensorflow-framework/src/test/java/org/tensorflow/framework/initializers/ConstantTest.java:
```diff
@@ -0,0 +1,146 @@
+/** Test the Constant initializer */
+public class ConstantTest {
+
+  private final TestSession.Mode tfMode = TestSession.Mode.EAGER;
```
Have these tests been run in graph mode too?

JimClarke5 replied:
I will change to run in both modes.

Craigacp reviewed (changes requested):
If I understand this correctly then the initializers are completely stateless, so they will give the same outputs for two layers if they are the same size and shape? I'm not sure that's ideal behaviour, but at least it's easy to replicate.

I am concerned that the default path seems to avoid setting the seed, and uses default seeds which are only present in the body of the method rather than flagged as part of the public API. Also I'd strongly prefer the easy path to be the one which requires a user to supply their own seed. That way when users try to ensemble together models (as currently appears to be best practice for the highest performance) then they know what to change to adjust the model starting point.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/Constant.java:
```diff
@@ -0,0 +1,81 @@
+      throw new IllegalArgumentException("DataType must be numeric or boolean: " + dtype.name());
+    }
+    if (this.numberValue != null) {
+      return tf.fill(dims, tf.dtypes.cast(tf.constant(numberValue.doubleValue()), dtype));
```
The basic requirement is to store either a number(integer + float) or a boolean.  What ever the most efficient way to do that is fine with me.

Craigacp replied:
Fair enough.

JimClarke5 replied:
I changed it to long longValue, double doubleValue, boolean booleanValue with an enum 
```
private enum ValueType {
    LONG, DOUBLE, BOOLEAN
 }
```
That way, the integer and double precision is maintained.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/Identity.java:
```diff
@@ -0,0 +1,96 @@
+
+    Operand<T> op;
+    Operand<T> zero = tf.dtypes.cast(tf.constant(0), dtype);
+    Operand<T> diag_ones =
```
OK

JimClarke5 replied:
Could you explain this a little more? Are you saying, add an op similar to `org.tensorflow.op.core.Zeros` (`tf.zeros`)? In TF Java, there is no `org.tensorflow.op.core.Ones` (nor `tf.ones`). In Python, `tf.ones` is implemented in python and ends up calling the `fill` op.

Craigacp replied:
I'd forgot there wasn't a `tf.ones`, but I think it would probably be worthwhile to add one. Either way I think generating an identity matrix is a useful operation and we shouldn't have it solely exist in the initializers package.

JimClarke5 replied:
I will open another PR on `tf.ones` and `tf.identity`. BTW, TF Python already has both.

JimClarke5 replied:
Java TensorFlow has `tf.identity`, but that is different than an Identity Matrix. Python TensorFlow has `tf.eye` (written in Python) for creating an Identity Matrix.

Craigacp replied:
Yeah I think the `tf.eye` mirrors numpy. Presumably `tf.identity` returns it's input?

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/utils/ShapeUtils.java:
```diff
@@ -0,0 +1,227 @@
+      @SuppressWarnings("unchecked")
+      Operand<TUint8> udims = (Operand<TUint8>) dims;
+        udims.asOutput().data().scalars().forEach(s -> result.add(s.getObject().longValue()));
+    }else { // shouldn't happen
```
OK

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/utils/ShapeUtils.java:
```diff
@@ -0,0 +1,227 @@
+                       session.runner().fetch(dims).run().get(0).expect(TUint8.DTYPE)) {
+            tensorResult.data().scalars().forEach(s -> result.add(s.getObject().longValue()));
+          }
+      }else { // shouldn't happen
```
OK

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/utils/ShapeUtils.java:
```diff
@@ -0,0 +1,227 @@
+  }
+
+  /**
+   * Determines if a shape is an unknown shape as provided in <cade>Shape.unknown()</code>.
```
OK

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/HeUniform.java:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @see <a href="https://www.cv-foundation.org/openaccess/content_iccv_2015/html/He_Delving_Deep_into_ICCV_2015_paper.html">He et al., 2015</a>
+ */
+public class HeUniform<T extends TType, U extends TNumber> extends VarianceScaling<T, U> {
```
Since we are not totally bound to the Keras, way, this makes sense to me.

Should we reconsider, LeCunNormal/LeCunUniform too?

JimClarke5 replied:
Also, should we restrict the distribution to `TRUNCATED_NORMAL` and `UNIFORM`, and show exception on `UNTRUNCATED_NORMAL`

Craigacp replied:
Erm, I think that allowing all three enum values seems reasonable (someone might want to try all of them and they aren't incorrect), and yes I'd collapse LeCun as well.

JimClarke5 replied:
Ok

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/Identity.java:
```diff
@@ -0,0 +1,96 @@
+  @Override
+  public Operand<T> call(Operand<TInt64> dims, DataType<T> dtype) {
+    if (!dtype.isFloating()) {
+      throw new IllegalArgumentException("DataType must be numeric or boolean: " + dtype.name());
```
You're right, it should only be a float.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/utils/ShapeUtils.java:
```diff
@@ -0,0 +1,227 @@
+   * <p>The compatibility relation is reflexive and symmetric, but not transitive. For example,
+   * <code>Shape(32, 784)</code> is compatible with <code>Shape.unknown()</code>, and <code>
+   * Shape.unknown()</code> is compatible with <code>Shape(4, 4)</code>, but <code>Shape(32, 784)
+   * </code> is not compatible with <code>Shape(4, 4)</code>.
```
No, it is not the same as broadcasting. Compatible shapes must  have the same number of dimensions and for each dimension pair, one dimension has to equal the other dimensions or at least one of the dimensions in the pair has to be `UNKNOWN_SIZE`.  
`Numpy` broadcasting allows different dimensions, but paired dimensions have to either be equal, or one dimension must be `1`. If one shape has less dimensions than another shape, the smaller shape is "stretched" with dimensions of `1`.

JimClarke5 replied:
I have upgraded the JavaDoc to explain the difference between compatible shapes and broadcastable shapes.

JimClarke5 replied:
Should `isCompatibleWith` be moved to `org.tensorflow.ndarray.Shape?`

Craigacp replied:
The javadoc change looks good, @karllessard do you think this method should be moved?

karllessard replied:
Probably yes, it would perfectly fit as an instance method of `Shape`: `shape.isCompatibleWith(...)`

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/Initializers.java:
```diff
@@ -0,0 +1,45 @@
+  HE_NORMAL(HeNormal::new),
+  HE_UNIFORM(HeUniform::new),
+  LECUN_NORMAL(LeCunNormal::new),
+  LECUN_UNIFORM(LeCunUniform::new);
```
Seeding should be true of any Initializer that uses a random generator. This is more of a convenience, so maybe we don't need this class in `framework`. Also the constructors for the the other Random initializers should then require a `seed`.

JimClarke5 replied:
I have changed a CTORs to require a `seed` parameter where a `seed` is used.

JimClarke5 commented:
The Initializers are stateless, if you don't count the properties used to instantiate them.  For a given `Initializer` instance, provided that you use the same shape and datatype, the outputs should be identical depending on the seed value for the ones that use random number generation. 

I am changing all the ctors to require seed for those initializers that use random number generation.

JimClarke5 commented:
I have updated the repository. Merged `LeCun`, `He`, and `Glorot` into one class each. Changed all the classes with a `seed` parameter to require a `seed` in the CTORs. Did JavaDoc clean up and formatting cleanup. Each test case now runs once under `EAGER` mode, and once under `GRAPH` mode.

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/Constant.java:
```diff
@@ -36,19 +36,37 @@
    *
    * @param tf the TensorFlow Ops
-   * @param value a number value
+   * @param value a double value
```
The javadoc says double, but the constructor expects a long.

JimClarke5 replied:
OK

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/Glorot.java:
```diff
@@ -0,0 +1,84 @@
+   * @param distribution The distribution type for the Glorot initializer.
+   * @param seed the seed for random number generation. An initializer created with a given seed
+   *     will always produce the same random tensor for a given shape and dtype.
+   * @throws java.lang.IllegalArgumentException if the distribution is not TRUNCATED_NORMAL or
```
So I understand that this is to replicate the behaviour in Keras, where GlorotNormal means use the Glorot variance and sample from a truncated normal, but why does it throw if people want to draw from a regular normal distribution? Also it looks like the super constructor doesn't throw IllegalArgumentException in this case, only if it got a zero `SCALE` param.

JimClarke5 replied:
In Keras, they only use Normal and Uniform. But it seems a little arbitrary. 
Here is the original discussion from the Keras thread on why they chose these two.
[Keras Issue on `GlorotNormal` vs `GlorotUniform` ](https://github.com/keras-team/keras/issues/52)

The `@throws` class was left in by mistake when I used to check for only TRUNCATED_NORMAL and UNIFORM. I will remove it.

Craigacp replied:
I think exposing all three is fine, and maybe a note to say that Keras uses TRUNCATED_NORMAL in the ones that produce normal values.

JimClarke5 replied:
I added class JavaDoc in all three types, `Glorot`, `He`, and `LeCun`, highlighting how to reproduce `xxxxxNormal`, and `xxxxxUniform` by using `TRUNCATED_NORMAL` and `UNIFORM`, respectively.

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/Orthogonal.java:
```diff
@@ -90,8 +72,9 @@ public Orthogonal(Ops tf, Long seed) {
    * @param tf the TensorFlow Ops
-   * @param gain the gain to be applied to the Matrix
-   * @param seed the seed for random number generation
+   * @param gain the gain to be applied to the Matrix. Default is 1.0.
```
This default notice should be on the other constructor as this one can't use defaults.

JimClarke5 replied:
OK

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/RandomNormal.java:
```diff
@@ -47,39 +49,33 @@
    * @param tf the TensorFlow Ops
-   * @param mean Mean of the random values to generate.
-   * @param stddev Standard deviation of the random values to generate.
+   * @param mean Mean of the random values to generate. Default is 0.
```
Default doc needs to move to the place where it's defaulted. Also I'd make it refer to the default value that way the comment can't become stale (e.g. `{@link MEAN_DEFAULT}`).

JimClarke5 replied:
OK

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/RandomUniform.java:
```diff
@@ -49,29 +51,23 @@
-   * @param minval Lower bound of the range of random values to generate (inclusive).
-   * @param maxval Upper bound of the range of random values to generate (exclusive).
-   * @param seed Used to create random seeds.
+   * @param minval Lower bound of the range of random values to generate (inclusive). Default is
```
Move the default doc.

JimClarke5 replied:
OK

JimClarke5 replied:
OK

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/TruncatedNormal.java:
```diff
@@ -49,29 +50,21 @@
-   * @param mean Mean of the random values to generate.
-   * @param stddev Standard deviation of the random values to generate.
-   * @param seed Used to create random seeds
+   * @param mean Mean of the random values to generate. Default is 0.
```
Default doc.

JimClarke5 replied:
OK

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/VarianceScaling.java:
```diff
@@ -188,14 +188,14 @@ public VarianceScaling(Ops tf, double scale, Mode mode, Distribution distributio
-  public static enum Distribution {
+  public enum Distribution {
     TRUNCATED_NORMAL,
     UNTRUNCATED_NORMAL,
```
I'd missed this last time, could this just be `NORMAL` not `UNTRUNCATED_NORMAL`?

JimClarke5 replied:
OK

Craigacp reviewed (changes requested):
Few small things with the Javadoc.

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/Orthogonal.java:
```diff
@@ -0,0 +1,139 @@
+    for (; i < dimsShape.numDimensions() - 1; i++) num_rows *= dimsShape.size(i);
+    long num_cols = dimsShape.size(i);
+    Shape flat_shape = Shape.of(Math.max(num_rows, num_cols), Math.min(num_rows, num_cols));
+    long lseed = this.seed == null ? 0L : this.seed;
```
This isn't necessary if we make seed be a `long` not a `Long`. I think we should prevent people from passing nulls to the constructor.

Craigacp reviewed (changes requested):
The seeds should be primitives not boxed primitives that'll let you elide the conditional logic. Either that or document that if the seed is null it defaults to zero (but I'd prefer them to be primitives). Other than that and a missing bit of Javadoc in VarianceScaling this looks good and is ready to be merged.

Craigacp reviewed (dismissed):
LGTM, thanks Jim.

karllessard commented on tensorflow-framework/src/main/java/org/tensorflow/framework/initializers/Identity.java:
```diff
@@ -0,0 +1,100 @@
+    }
+    boolean isSquare = shape.size(0) == shape.size(1);
+    long diag_size = Math.min(shape.size(0), shape.size(1));
+    Shape diagShape = Shape.of(diag_size);
```
There is still of few variables like this one in the PR using snake_case instead of camelCase.

karllessard reviewed (commented):
I think we all agree that in general, this PR is ready to be merged. Up to you @JimClarke5 to update now the small details we've mentioned or do it later (I like though the idea of moving `isCompatibleWith` to the `ndarray` library).

karllessard commented:
Another thing, I have some concerns about the naming the some classes, like `Constant`, which can cause conflict with the [`Constant`](https://github.com/tensorflow/java/blob/master/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/Constant.java) in the core API, forcing us to use the canonical names in case both types are present in some class. 

I understand though that if we prefix/suffix this class, we should do it for all of our initializers as well (and what about our current optimizers?). So maybe we should just find another name for these?

Note that we are never protected from conflicts with the ops classes since most of them are generated. For example one day, a `Ones` kernel might appear as well...

JimClarke5 commented:
I have fixed the snake_case issues.

The names are the same ones used in Keras. IMO, most users will use `tf.constant()` and not the ops `Constant` class directly.

Craigacp commented:
> I have fixed the snake_case issues.
> 
> The names are the same ones used in Keras. IMO, most users will use `tf.constant()` and not the ops `Constant` class directly.

Yeah, but if they assign `tf.constant()` to a variable they'll need to bring the `Constant` type into scope unless they are using Java 10+ and use `var`. So it'll conflict.

JimClarke5 commented:
It's besides the point but Shape already convicts with `org/tensorflow.op.core.Shape `and `org.tensorflow.ndarray.Shape`.

If you want, I can rename them all to `xxxxxInitialzier`.

JimClarke5 commented:
I have moved `isCompatibleWith` to ndarray `Shape`, and added test for it in `ShapeTest`.

karllessard commented:
> > I have fixed the snake_case issues.
> > The names are the same ones used in Keras. IMO, most users will use `tf.constant()` and not the ops `Constant` class directly.
> 
> Yeah, but if they assign `tf.constant()` to a variable they'll need to bring the `Constant` type into scope unless they are using Java 10+ and use `var`. So it'll conflict.

Exactly the case I had in mind as well. They can just keep a reference to `Operand` and it will be fine but I personally like to preserve the original type of the ops as I add it to the graph. But like you said, in Java 10+ and Kotlin it will be less frequent. So let’s keep the current naming strategy for now and review it later if needed.

karllessard commented:
So this PR is good to go, thanks @JimClarke5 ! Please review the conflict raised above and I’ll merge it after the release of 0.2.0 (which I plan to do this weekend)

JimClarke5 commented:
The conflict in `Shape` had to do with JavaDoc indenting. As I am using the Google Java formatter in InteliliJ, I resolved he conflict with my version.

Did we want to rename `Constant` to `ConstantInitializer`, `Zeros` to `ZerosInitializer` and `Ones` to `OnesInitializer` ?
I noticed that in TF Python, 2.3.1, there are now initializer methods defined in `init_ops_v2.py` directly under the `tensorflow/python/ops` directory (outside of Keras), called `tf.constant_initializer`, presumably to avoid the conflict with `tf.constant`, similar to our predicament. They also have` tf.ones_initializer`,` tf.zeros_initializer`, among others.  They use the same Class names as we currently have, but the method aliases are all in the form `tf.xxxx_initializer`.

karllessard commented:
If we do it for the initializers, we should do it for the optimizers as well for consistency. That is why I think it can be out of scope of this PR to make that change and we can do it later, wdyt?

JimClarke5 commented:
I vote we leave it as is for now. We still have losses and metrics to go, so by then we can get a holistic view how to handle this issue.

karllessard commented:
@JimClarke5 , I’m sorry to tell you this but it looks like your PR has new conflicts since I’ve updated the master branch for the next development iteration. Can you please fix them so we can merge it?

karllessard commented:
Ok that’s fine, it just had trouble to rebase but a simple merge did the trick, thanks for your great work!

JimClarke5 commented:
So I don’t have to fix anything, right.

karllessard commented:
Nope, it’s already merged! I had to do a few quick fixes in the javadoc though to be able to deploy this new snapshot, we should enable these lint checks locally so a developer can catch the errors upfront, there is a story about it: #7
